source/game_sa/CarCtrl.cpp

```diff
@@ -320,37 +367,85 @@ void CCarCtrl::FindNodesThisCarIsNearestTo(CVehicle* vehicle, CNodeAddress& node
+    return plugin::CallAndReturn<int8, 0x422090, CNodeAddress, CNodeAddress, CNodeAddress, bool*>(oldNode, currentNode, newNode, bSharpTurn);
 }
 
 // 0x422620
 float CCarCtrl::FindPercDependingOnDistToLink(CVehicle* vehicle, CCarPathLinkAddress linkAddress) {
-    return plugin::CallAndReturn<float, 0x422620, CVehicle*, CCarPathLinkAddress>(vehicle, linkAddress);
+    const auto& pathLinkPos = ThePaths.GetCarPathLink(linkAddress).GetNodeCoors();
+    const float dist        = (pathLinkPos - vehicle->GetPosition2D()).Magnitude();
+    if (dist < 5.0f) {
+        return 0.5f;
+    }
+    if (dist >= 15.0f) {
+        return 1.0f;
+    }
+    return (dist - 5.0f) * 0.05f + 0.5f;
 }
 
 // 0x421770
-int32 CCarCtrl::FindSequenceElement(int32 arg1) {
-    return plugin::CallAndReturn<int32, 0x421770, int32>(arg1);
+int32 CCarCtrl::FindSequenceElement(int32 element) {
+    if (bSequenceOtherWay) {
+        return (SequenceRandomOffset + element) % SequenceElements;
+    } else {
+        return (SequenceRandomOffset + SequenceElements - element) % SequenceElements;
+    }
 }
 
 // 0x4224E0
-float CCarCtrl::FindSpeedMultiplier(float arg1, float arg2, float arg3, float arg4) {
-    return plugin::CallAndReturn<float, 0x4224E0, float, float, float, float>(arg1, arg2, arg3, arg4);
+float CCarCtrl::FindSpeedMultiplier(float angle, float maxAngle, float minAngle, float minSpeed) {
+    angle = CGeneral::LimitRadianAngle(angle);
+    angle = std::abs(angle);
+
+    float angleDiff = angle - maxAngle;
+    if (angleDiff < 0.0f) {
+        angleDiff = 0.0f;
+    }
+
+    float range = minAngle - maxAngle;
+    if (angleDiff > range) {
+        return minSpeed;
+    }
+
+    return std::lerp(1.0f, minSpeed, angleDiff / range);
 }
 
 // 0x424130
-float CCarCtrl::FindSpeedMultiplierWithSpeedFromNodes(int8 arg1) {
-    return plugin::CallAndReturn<float, 0x424130, int8>(arg1);
+float CCarCtrl::FindSpeedMultiplierWithSpeedFromNodes(int8 speedFromNodes) {
+    switch (speedFromNodes) {
+    case -1: return 0.5f;
+    case 0:  return 0.65f;
+    case 1:  return 1.0f;
+    case 2:  return 2.3f;
+    }
+    return 1.0f;
 }
 
+// 0x422370
 float CCarCtrl::FindGhostRoadHeight(CVehicle* vehicle) {
     return plugin::CallAndReturn<float, 0x422370, CVehicle*>(vehicle);
 }
 
 // 0x42B270
-void CCarCtrl::FireHeliRocketsAtTarget(CAutomobile* entityLauncher, CEntity* entity) {
-    plugin::Call<0x42B270, CAutomobile*, CEntity*>(entityLauncher, entity);
+void CCarCtrl::FireHeliRocketsAtTarget(CHeli* heli, CEntity* target) {
+    if ((heli->m_nVehicleWeaponInUse != CAR_WEAPON_DOUBLE_ROCKET && heli->m_nVehicleWeaponInUse != CAR_WEAPON_NOT_USED)
+        || CTimer::m_snTimeInMilliseconds / 250 == CTimer::m_snPreviousTimeInMilliseconds / 250) {
+        return;
+    }
+
+    const auto& heliPos   = heli->GetPosition();
+    const auto& targetPos = target->GetPosition();
+
+    if ((heliPos - targetPos).Magnitude() >= 80.0f) {
+        return;
+    }
```

Comment:
prefer `SquaredMagnitude` over `Magnitude` wherever possible
```suggestion
    if ((heliPos - targetPos).SquaredMagnitude() >= sq(80.0f)) {
```